graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt

```diff
@@ -257,87 +247,83 @@ class DgsSchemaProvider(
-                                    it
-                                } else {
-                                    it.joinToString()
-                                }
+                    args.add(convertValue)
+                }
+
+                parameter.isAnnotationPresent(RequestHeader::class.java) -> {
+                    val annotation = parameter.getAnnotation(RequestHeader::class.java)
+                    val parameterName = annotation.value.ifBlank { parameterNames[idx] }
+
+                    args.add(
+                        DgsContext.getRequestData(environment)?.headers?.get(parameterName)?.let {
+                            if (parameter.type.isAssignableFrom(List::class.java)) {
+                                it
+                            } else {
+                                it.joinToString()
                             }
-                        )
-                    }
+                        }
+                    )
+                }
 
-                    environment.containsArgument(parameterNames[idx]) -> {
-                        val parameterValue: Any = environment.getArgument(parameterNames[idx])
-                        val convertValue = objectMapper.convertValue(parameterValue, parameter.type)
-                        args.add(convertValue)
-                    }
+                environment.containsArgument(parameterNames[idx]) -> {
+                    val parameterValue: Any = environment.getArgument(parameterNames[idx])
+                    val convertValue = objectMapper.convertValue(parameterValue, parameter.type)
+                    args.add(convertValue)
+                }
 
-                    parameter.type == DataFetchingEnvironment::class.java || parameter.type == DgsDataFetchingEnvironment::class.java -> {
-                        args.add(environment)
-                    }
-                    else -> {
-                        logger.warn("Unknown argument '${parameterNames[idx]}' on data fetcher ${dgsComponent.javaClass.name}.${method.name}")
-                        // This might cause an exception, but parameter's the best effort we can do
-                        args.add(null)
-                    }
+                parameter.type == DataFetchingEnvironment::class.java || parameter.type == DgsDataFetchingEnvironment::class.java -> {
+                    args.add(environment)
+                }
+                else -> {
+                    logger.warn("Unknown argument '${parameterNames[idx]}' on data fetcher ${dgsComponent.javaClass.name}.${method.name}")
+                    // This might cause an exception, but parameter's the best effort we can do
+                    args.add(null)
                 }
             }
-
-            method.invoke(dgsComponent, *args.toTypedArray())
-        } catch (ex: InvocationTargetException) {
-            throw ex.targetException
         }
+
+        return ReflectionUtils.invokeMethod(method, dgsComponent, *args.toTypedArray())
```

Comment:
This handles InvocationTargetException, rethrowing targetException, as well as other cases.